### PR TITLE
[CARBONDATA-3243] Updated DOC for No-Sort Compaction and a few Fixes

### PR DESCRIPTION
**Pull request by NamanRastogi**
1. Updated Doc
2. Checking **_SORT_SCOPE_** in session property **_CARBON.TABLE.LOAD.SORT.SCOPE_** in **_CarbonTable.getSortScope()_**
3. Changed default Sort Scope in **_SortScopeOptions.getSortScope()_**
4. Validation for Load Option **_SORT_SCOPE_**
5. Fixed a mistake introduced in [PR 3029](https://github.com/apache/carbondata/pull/3029) in **_SingleThreadFinalSortFilesMerger.addInMemoryRawResultIterator()_**. Add the iterator in priority queue only of record is found in iterator, else not.

 - [x] Any interfaces changed?  ->  NO
 - [x] Any backward compatibility impacted?  ->  NO
 - [x] Document update required?   ->  NO
 - [x] Testing done  -> Yes
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2472/

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/execution/command/CarbonHiveCommands.scala [hunk 1]
```diff
@@ -127,6 +127,9 @@ object CarbonSetCommand {
     else if (isCarbonProperty) {
       sessionParams.addProperty(key, value)
     }
+    else {
```
[the remark this answers is not in the document]
Comment (NamanRastogi) [reply integration/spark2/src/main/scala/org/apache/spark/sql/hive/execution/command/CarbonHiveCommands.scala]: Done.

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2473/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2254/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10512/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2255/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2474/

File: docs/configuration-parameters.md [hunk 2]
```diff
@@ -208,6 +208,7 @@ RESET
 | carbon.options.date.format                 | Specifies the data format of the date columns in the data being loaded |
 | carbon.options.timestamp.format            | Specifies the timestamp format of the time stamp columns in the data being loaded |
 | carbon.options.sort.scope                 | Specifies how the current data load should be sorted with. **NOTE:** Refer to [Data Loading Configuration](#data-loading-configuration)#carbon.sort.scope for detailed information. |
+| carbon.table.load.sort.scope              | Overrides the SORT_SCOPE provides in CREATE TABLE.                         |
```
Comment (qiuchenjian) [inline docs/configuration-parameters.md]:
```suggestion
| carbon.table.load.sort.scope              | Override the SORT_SCOPE provided in CREATE TABLE.                         |
```

File: docs/dml-of-carbondata.md [hunk 3]
```diff
@@ -49,6 +49,7 @@ CarbonData DML statements are documented here,which includes:
 | [COMMENTCHAR](#commentchar)                             | Character used to comment the rows in the input csv file. Those rows will be skipped from processing |
 | [HEADER](#header)                                       | Whether the input csv files have header row                  |
 | [FILEHEADER](#fileheader)                               | If header is not present in the input csv, what is the column names to be used for data read from input csv |
+| [SORT_SCOPE](#sort_scope)                               | Sort Scope to be used for current load.                      |
```
Comment (qiuchenjian) [inline docs/dml-of-carbondata.md]:
```suggestion
| [SORT_SCOPE](#sort_scope)                               | Sort Scope is used for current load.                      |
```

File: docs/dml-of-carbondata.md [hunk 4]
```diff
@@ -106,6 +107,13 @@ CarbonData DML statements are documented here,which includes:
     OPTIONS('FILEHEADER'='column1,column2') 
     ```
 
+  - ##### SORT_SCOPE:
+  Sort Scope to be used for the current load. This overrides the Sort Scope of Table.
```
Comment (qiuchenjian) [inline docs/dml-of-carbondata.md]:
```suggestion
  Sort Scope is used for the current load. This overrides the Sort Scope of Table.
```

File: docs/configuration-parameters.md (hunk 2 above)
Comment (NamanRastogi) [reply docs/configuration-parameters.md]: "provides" changed to "provided". This was a spelling mistake.
"Overrides" is correct. No Change.

File: docs/dml-of-carbondata.md (hunk 3 above)
Comment (NamanRastogi) [reply docs/dml-of-carbondata.md]: "to be" is correct. No change.

File: docs/dml-of-carbondata.md (hunk 4 above)
Comment (NamanRastogi) [reply docs/dml-of-carbondata.md]: "to be" is correct. No change.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2263/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2267/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10525/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2486/

File: integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala [hunk 5]
```diff
@@ -1201,6 +1202,17 @@ abstract class CarbonDDLSqlParser extends AbstractCarbonSparkSQLParser {
...
+    if(options.exists(_._1.equalsIgnoreCase("SORT_SCOPE"))) {
+      val optionValue: String = options.get("sort_scope").get.head._2
+      if (!CarbonUtil.isValidSortOption(optionValue)) {
+        throw new InvalidConfigurationException(
+          s"Passing invalid SORT_SCOPE '$optionValue', valid SORT_SCOPE are 'NO_SORT'," +
+          s" 'BATCH_SORT', 'LOCAL_SORT' and 'GLOBAL_SORT' ")
+      }
+
+    }
+
```
Comment (shardul-cr7) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala]: Remove empty lines and properly format the code.

Comment (NamanRastogi) [reply integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala]: Done.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2274/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10533/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2494/

File: processing/src/main/java/org/apache/carbondata/processing/sort/sortdata/SingleThreadFinalSortFilesMerger.java [hunk 6]
```diff
@@ -135,8 +135,8 @@ public void addInMemoryRawResultIterator(List<RawResultIterator> sortedRawResult
               noDicAndComplexColumns, sortParameters, measureDataType);
       if (inMemorySortTempChunkHolder.hasNext()) {
         inMemorySortTempChunkHolder.readRow();
+        recordHolderHeapLocal.add(inMemorySortTempChunkHolder);
```
[the remark this answers is not in the document]
Comment (NamanRastogi) [reply processing/src/main/java/org/apache/carbondata/processing/sort/sortdata/SingleThreadFinalSortFilesMerger.java]: `inMemorySortTempChunkHolder`  ->  adapter for RawResultIterator on sorted files
`recordHolderHeapLocal`  ->  Priority Queue

This line adds `inMemorySortTempChunkHolder` to the `recordHolderHeapLocal` so that it can be used for final merge sorting.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2325/

**Comment by xuchuanyin** [comment]
Since the doc for no-sort compaction is removed, please optimize the title and description of the PR correspondingly.

**Comment by xuchuanyin** [comment]
Can other committers check this PR for the bug fix?
I'm not sure about the impaction since previously we didn't encounter error in data loading.

@NamanRastogi Can you describe the problem you are facing that related to these bug fix?

**Comment by NamanRastogi** [comment]
@xuchuanyin I have removed the Explanation of Compaction Performance. But Document update is still required for `SET ALTER.TABLE.LOAD.SORT.SCOPE.<database>.<table>` and Sort-Scope in `LOAD DATA` command.

Problems:
1. **SortScopeOptions.getSortScope**
    Default Sort Scope after PR is NO_SORT. So default here also should be Default Sort Scope, not some hardcoded value.

2. **CarbonTable.getSortScope**
    After [PR 3014](https://github.com/apache/carbondata/pull/3014), session property for table level sort scope is added. This overrides the table level Sort Scope provided in `CREATE TABLE` command. CarbonTable.getSortScope should also reflect this behavior.

3. **CarbonDDLSqlParser.validateOptions**
    The Sort Scope for Load Data command was not getting verified.

4. **SingleThreadFinalSortFilesMerger.getSortScope**
    Object `inMemorySortTempChunkHolder` should be added to the `recordHolderHeapLocal` Only if a row is found in `inMemorySortTempChunkHolder`, else not.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2374/

File: docs/dml-of-carbondata.md [hunk 7]
```diff
@@ -458,4 +466,4 @@ CarbonData DML statements are documented here,which includes:
   ```
   CLEAN FILES FOR TABLE carbon_table
   ```
-
+
```
Comment (kumarvishal09) [inline docs/dml-of-carbondata.md]: revert this change

File: processing/src/main/java/org/apache/carbondata/processing/sort/sortdata/SingleThreadFinalSortFilesMerger.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (xuchuanyin) [reply processing/src/main/java/org/apache/carbondata/processing/sort/sortdata/SingleThreadFinalSortFilesMerger.java]: still cannot understand this change?
Is it for a bug?
If yes, please mention that in the PR description about the scenario and the root cause and add test case for that. -- I'd recommend you to raise a new PR for it.

[the remark this answers is not in the document]
Comment (NamanRastogi) [reply processing/src/main/java/org/apache/carbondata/processing/sort/sortdata/SingleThreadFinalSortFilesMerger.java]: Yes. It is for a bug. Add the `inMemorySortTempChunkHolder` object in the priority queue only if data is present in that object. Earlier it was adding it anyway. This fixes that mistake.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2598/

**Comment by xuchuanyin** [comment]
@NamanRastogi 
As we talked, it's not a problem before PR3029. So you'd better to explain that this change is to revert the mistake introduced by PR3029. By saying this, the others can just ignore this change of line while they understand this PR. Please add this to the PR DESCRIPTION.

Others LGTM.

**Comment by xuchuanyin** [comment]
@NamanRastogi 
"Fixed a bug in SingleThreadFinalSortFilesMerger.addInMemoryRawResultIterator. Add the iterator in priority queue only of record is found in iterator, else not."
===
Please do not tell it is a bug. Just say ‘it is for the mistake introduced by PR3029, here I just revert that mistake.’

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10633/

File: docs/dml-of-carbondata.md (hunk 7 above)
Comment (NamanRastogi) [reply docs/dml-of-carbondata.md]: Done.

**Comment by NamanRastogi** [comment]
@xuchuanyin Updated PR Description.

**Comment by xuchuanyin** [comment]
LGTM

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2601/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10636/

**Comment by kunal642** [comment]
LGTM
